src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -486,6 +486,68 @@ public String getCookie() {
             return getDisplayName();
         }
 
+        static String findThreadName(FlowNode flowNode){
+            ThreadNameAction tna = flowNode.getPersistentAction(ThreadNameAction.class);
+
+            if (tna != null) {
+                return tna.getThreadName();
+            }
+            return null;
+        }
+
+        static String findStageName(FlowNode flowNode){
+            LabelAction la = flowNode.getPersistentAction(LabelAction.class);
+
+            if (la != null) {
+                return la.getDisplayName();
+            }
+            return null;
+        }
+
+        /**
+         * Similar to {@link #getEnclosingLabel()}.
+         * However instead of returning the innermost label including labels inside node blocks this one
+         * concatenates all labels found outside the current (node) block
+         * */
+        private @CheckForNull String concatenateAllEnclosingLabels() {
+            if (!context.isReady()) {
+                return null;
+            }
+            FlowNode executorStepNode;
+            try (Timeout t = Timeout.limit(100, TimeUnit.MILLISECONDS)) {
+                executorStepNode = context.get(FlowNode.class);
+            } catch (Exception x) {
+                LOGGER.log(Level.FINE, null, x);
+                return null;
+            }
+
+            StringBuilder labelName = new StringBuilder();
+            for(FlowNode node: executorStepNode.getEnclosingBlocks()) {
+                String threadName = findThreadName(node);
+                String stageName = findStageName(node);
+                String currentLabelName = threadName != null ? threadName: stageName;
+                if (threadName != null && stageName != null) {
+                    LOGGER.log(FINE, "\"findLabelAndParseLog(): Got both thread: {0} and stage: $stageName. Will take thread name ({0}).", threadName);
+                }
+                if (currentLabelName != null) {
+                    labelName.append("#");
+                    labelName.append(currentLabelName);
+                }
+            }
+            return labelName.toString();
+        }
+
+        @Override
+        public String getAffinityKey() {
+            String enclosingLabels = concatenateAllEnclosingLabels();
+            String ownerTaskName = getOwnerTask().getName();
+            if (enclosingLabels != null) {
+                return ownerTaskName + enclosingLabels;
```

Comment:
To make sure I understand, what would the affinity key be for the node step in the following Pipeline, assuming the WorkflowJob is named `Test`:

    stage('Stage1') {
        parallel 'Branch1': {
            node ('Agent1') {
            }
        }, 'Branch2': {
        }
    }

`TestBranch1#Stage1`? `TestStage1#Branch1`? Something else entirely?

Reply:
This depends on the implementation of `getEnclosingBlocks()`. Currently it would return the inner blocks first - in that case that should be: `TestBranch1#Stage1`. However from functionality point of view it doesn't really matter. What is important is to have all of the block names & the job name in the key. Maybe for consistency I should either remove the '#' or add it between the job name and first enclosing block? I'll have some thought's on this one.

Reply:
Yeah, I agree that the specific format doesn't really matter as long as everything is in it, was mostly just curious. I initially thought it might be important to put a `#` between each enclosing label and the name to avoid unintentional matches, but I can't actually think of any cases that would cause problems, so it seems fine to me.